prow/istio-pilot-multicluster-e2e.sh

```diff
@@ -14,14 +14,108 @@
-export CLUSTERREG_DIR
+set -eux
 
-echo 'Running pilot multi-cluster e2e tests'
-date
-./prow/e2e-suite.sh --timeout 90m --cluster_registry_dir="$CLUSTERREG_DIR" --single_test e2e_pilotv2_v1alpha3 "$@"
+function setup_clusterreg () {
+    MAIN_CONFIG=""
+    for context in "${CLUSTERREG_DIR}"/*; do
+        if [[ -z "${MAIN_CONFIG}" ]]; then
+            MAIN_CONFIG="${context}"
+        fi
+        export KUBECONFIG="${context}"
+        kubectl delete ns istio-system-multi --ignore-not-found
+        kubectl delete clusterrolebinding istio-multi-test --ignore-not-found
+        kubectl create ns istio-system-multi
+        kubectl create sa istio-multi-test -n istio-system-multi
+        kubectl create clusterrolebinding istio-multi-test --clusterrole=cluster-admin --serviceaccount=istio-system-multi:istio-multi-test
+        CLUSTER_NAME=$(kubectl config view --minify=true -o "jsonpath={.clusters[].name}")
+        gen_kubeconf_from_sa istio-multi-test "${context}"
+    done
+    export KUBECONFIG="${MAIN_CONFIG}"
+}
+
+function gen_kubeconf_from_sa () {
+    local service_account=$1
+    local filename=$2
+
+    SERVER=$(kubectl config view --minify=true -o "jsonpath={.clusters[].cluster.server}")
+    SECRET_NAME=$(kubectl get sa "${service_account}" -n istio-system-multi -o jsonpath='{.secrets[].name}')
+    CA_DATA=$(kubectl get secret "${SECRET_NAME}" -n istio-system-multi -o "jsonpath={.data['ca\\.crt']}")
+    TOKEN=$(kubectl get secret "${SECRET_NAME}" -n istio-system-multi -o "jsonpath={.data['token']}" | base64 --decode)
+
+    cat <<EOF > "${filename}"
+      apiVersion: v1
+      clusters:
+         - cluster:
+             certificate-authority-data: ${CA_DATA}
+             server: ${SERVER}
+           name: ${CLUSTER_NAME}
+      contexts:
+         - context:
+             cluster: ${CLUSTER_NAME}
+             user: ${CLUSTER_NAME}
+           name: ${CLUSTER_NAME}
+      current-context: ${CLUSTER_NAME}
+      kind: Config
+      preferences: {}
+      users:
+         - name: ${CLUSTER_NAME}
+           user:
+             token: ${TOKEN}
+EOF
+}
+
+SUBNET_A="10.10.0.0/16"
+SUBNET_B="10.20.0.0/16"
+
+CONFIG_A=$(cat <<EOF
+kind: Cluster
+apiVersion: kind.sigs.k8s.io/v1alpha3
+networking:
+  podSubnet: ${SUBNET_A}
+  serviceSubnet: 10.255.10.0/24
+EOF
+)
+CONFIG_B=$(cat <<EOF
+kind: Cluster
+apiVersion: kind.sigs.k8s.io/v1alpha3
+networking:
+  podSubnet: ${SUBNET_B}
+  serviceSubnet: 10.255.10.0/24
+EOF
+)
+
+# Create two clusters. Explicitly delcare the subnet so we can connect the two later
+kind create cluster --name a --config <(echo "${CONFIG_A}")
+kind create cluster --name b --config <(echo "${CONFIG_B}")
+
+export HUB=istio-testing
+export TAG=istio-testing
+# Build just the images needed for the tests
+for image in pilot proxyv2 proxy_init app test_policybackend mixer citadel galley sidecar_injector kubectl node-agent-k8s; do
+   make docker.${image}
+done
```

Comment:
I'm surprised we have to do this in here.  We can't use images from a build stage, ie. via `docker pull ...`?

Reply:
Prow doesn't support pipelines/ordering of tests, so each one has to build everything.. it's very wasteful